### PR TITLE
Add support for BFRuntime gRPC server in stratum_bf

### PR DESCRIPTION
**Pull request by bocon13**
depends on #374

### CODE REVIEW

File: stratum/hal/bin/barefoot/README.run.md [hunk 1]
```diff
@@ -249,6 +249,22 @@ will configure device port 132 in 100G mode with Reed-Solomon (RS) FEC.
...
+customize this with the `-incompatible_bfrt_grpc_server_addr` flag`.
+For example to listen to all connection on TCP port 12345, use:
+
+```bash
+start-stratum.sh \
+-incompatible_enable_bfrt_grpc_server \
+-incompatible_bfrt_grpc_server_addr=0.0.0.0:12345
+```
+
 -----
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/README.run.md:268]: Cross reference the build readme to mention how to enable it

File: stratum/hal/lib/barefoot/BUILD [hunk 2]
```diff
@@ -229,3 +229,54 @@ cc_proto_library(
...
+        "//stratum/glue/net_util:ports",
+        "@com_github_gflags_gflags//:gflags",
+        "@com_github_grpc_grpc//:grpc++",
+        "@com_google_googletest//:gtest_main",
+    ] + select({
+        "with_bfrt_grpc_server": [
+            # PI needed for linking libdriver.so symbols
+            "//stratum/hal/lib/pi:pi_node_bf",
+        ],
+	    "//conditions:default": [],
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/BUILD:275]: spacing

File: stratum/hal/lib/barefoot/BUILD [hunk 3]
```diff
@@ -229,3 +229,54 @@ cc_proto_library(
...
+    }),
+    defines = select({
+        "with_bfrt_grpc_server": ["WITH_BFRT_GRPC_SERVER"],
+	    "//conditions:default": [],
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/BUILD:279]: spacing

File: bazel/external/bfsde.BUILD [hunk 4]
```diff
@@ -28,37 +34,106 @@ cc_library(
...
+        "@com_github_grpc_grpc//:grpc++",
+        "@com_google_googleapis//google/rpc:code_cc_proto",
+        "@com_google_googleapis//google/rpc:status_cc_proto",
+    ],
+)
+
+proto_library(
+    name = "bfruntime_proto",
+    srcs = ["barefoot-bin/src/bf_rt/proto/bfruntime.proto"],
+    # FIXME(bocon): there's a bug in the cc_grpc_library rule that doesn't like this
```
Comment (bocon13) [inline bazel/external/bfsde.BUILD:98]: Downgrade to TODO

File: stratum/hal/lib/barefoot/BUILD [hunk 5]
```diff
@@ -229,3 +229,54 @@ cc_proto_library(
...
+    deps = [
+        ":bfrt_grpc_server",
+        "//stratum/glue/net_util:ports",
+        "@com_github_gflags_gflags//:gflags",
+        "@com_github_grpc_grpc//:grpc++",
+        "@com_google_googletest//:gtest_main",
+    ] + select({
+        "with_bfrt_grpc_server": [
+            # PI needed for linking libdriver.so symbols
+            "//stratum/hal/lib/pi:pi_node_bf",
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/BUILD:273]: use `:pi_bf` from #375

File: bazel/external/bfsde.BUILD [hunk 6]
```diff
@@ -11,14 +17,14 @@ package(
...
-        "barefoot-bin/lib/libpython3.4m.so",
-    ],
+    srcs = glob([
+        "barefoot-bin/lib/libavago.so*",
+        "barefoot-bin/lib/libbf_switchd_lib.so*",
+        "barefoot-bin/lib/libbfsys.so*",
+        "barefoot-bin/lib/libbfutils.so*",
+        "barefoot-bin/lib/libdriver.so*",
+        "barefoot-bin/lib/libpython3.4m.so*",
+    ]),
```
Comment (pudelkoM) [inline bazel/external/bfsde.BUILD:27]: This change should be one PR

File: bazel/rules/package_rule.bzl [hunk 7]
```diff
@@ -30,8 +30,9 @@ def pkg_tar_with_symlinks(
         cmd = cmd + """
             dir=%s
             for f in $(SRCS); do
-                symlink_target=$$(readlink $$f | xargs -I{} find {} -type l | xargs readlink || true)
-                if [[ -n $$symlink_target ]]; then
+                symlink=$$(readlink "$$f" | xargs -I{} find {} -type l)
+                if [[ -n "$$symlink" ]]; then
+                    symlink_target=$$(readlink "$$symlink")
```
Comment (pudelkoM) [inline bazel/rules/package_rule.bzl:35]: This change should be one PR

File: stratum/hal/lib/barefoot/bfrt_grpc_server.h [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef STRATUM_HAL_LIB_BAREFOOT_BFRT_GRPC_SERVER_H_
+#define STRATUM_HAL_LIB_BAREFOOT_BFRT_GRPC_SERVER_H_
+
+namespace stratum {
+namespace hal {
+namespace barefoot {
+
+void start_bfrt_server_if_enabled();
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bfrt_grpc_server.h:12]: Pascal case

File: bazel/rules/package_rule.bzl (hunk 7 above)
Comment (bocon13) [reply bazel/rules/package_rule.bzl:35]: moved to #378

File: bazel/external/bfsde.BUILD (hunk 6 above)
Comment (bocon13) [reply bazel/external/bfsde.BUILD:27]: moved to #374

**Comment by bocon13** [comment]
superseded by #388

**Pull request closed without merging**